File: src/sugtao4423/lod/utils/DBUtil.java
package sugtao4423.lod.utils;

import java.util.ArrayList;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import sugtao4423.lod.AccountDB;
import sugtao4423.lod.Keys;
import sugtao4423.lod.dataclass.Account;

public class DBUtil{

	private SQLiteDatabase db;

	public DBUtil(Context context){
		db = new AccountDB(context).getWritableDatabase();
	}

	public Account getAccount(String screenName){
		Cursor c = db.rawQuery(String.format("SELECT * FROM accounts WHERE %s='%s'", Keys.SCREEN_NAME, screenName), null);
		if(!c.moveToFirst()){
			return null;
		}
		Account account = getAccount(c);
		c.close();
		return account;
	}

	public Account[] getAccounts(){
		Cursor c = db.rawQuery("SELECT * FROM accounts", null);
		ArrayList<Account> accountList = new ArrayList<Account>();

		boolean mov = c.moveToFirst();
		while(mov){
			accountList.add(getAccount(c));
			mov = c.moveToNext();
		}
		c.close();
		return (Account[])accountList.toArray(new Account[0]);
	}

	private Account getAccount(Cursor c){
		String screen_name = c.getString(0);
		String ck = c.getString(1);
		String cs = c.getString(2);
		String at = c.getString(3);
		String ats = c.getString(4);
		long listAsTL = c.getLong(5);
		int autoLoadTLInterval = c.getInt(6);
		String listIds = c.getString(7);
		String listNames = c.getString(8);
		String appLoadLists = c.getString(9);

		long[] selectListIds;
		if(listIds.equals("")){
			selectListIds = new long[0];
		}else{
			String[] strarr = listIds.split(",");
			selectListIds = new long[strarr.length];
			for(int i = 0; i < strarr.length; i++){
				selectListIds[i] = Long.parseLong(strarr[i]);
			}
		}
		
		String[] selectListNames = listNames.equals("") ? new String[0] : listNames.split(",");
		String[] startAppLoadLists = appLoadLists.equals("") ? new String[0] : appLoadLists.split(",");

		return new Account(screen_name, ck, cs, at, ats, listAsTL, autoLoadTLInterval, selectListIds, selectListNames, startAppLoadLists);
	}

	public boolean existsAccount(String screenName){
		Cursor c = db.rawQuery(String.format("SELECT COUNT(*) FROM accounts WHERE %s='%s'", Keys.SCREEN_NAME, screenName), null);
		c.moveToFirst();
		int count = c.getInt(0);
		c.close();
		return count > 0;
	}

	public void addAcount(Account account){
		db.execSQL(String.format("INSERT INTO accounts VALUES('%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s')",
				account.getScreenName(), account.getConsumerKey(), account.getConsumerSecret(),
				account.getAccessToken(), account.getAccessTokenSecret(), account.getListAsTL(), account.getAutoLoadTLInterval(),
				Utils.implode(account.getSelectListIds()),
				Utils.implode(account.getSelectListNames()),
				Utils.implode(account.getStartAppLoadLists())));
	}

	public void deleteAccount(Account account){
		db.execSQL(String.format("DELETE FROM accounts WHERE %s='%s'", Keys.SCREEN_NAME, account.getScreenName()));
	}

	public void updateListAsTL(long listAsTL, String screenName){
		db.execSQL(getUpdate1ColumnFromEq1Column(Keys.LIST_AS_TIMELINE, listAsTL, screenName));
	}

	public void updateAutoLoadTLInterval(int autoLoadTLInterval, String screenName){
		db.execSQL(getUpdate1ColumnFromEq1Column(Keys.AUTO_LOAD_TL_INTERVAL, autoLoadTLInterval, screenName));
	}

	public void updateSelectListIds(String ids, String screenName){
		db.execSQL(getUpdate1ColumnFromEq1Column(Keys.SELECT_LIST_IDS, ids, screenName));
	}

	public void updateSelectListNames(String names, String screenName){
		db.execSQL(getUpdate1ColumnFromEq1Column(Keys.SELECT_LIST_NAMES, names, screenName));
	}

	public void updateStartAppLoadLists(String startAppLoadLists, String screenName){
		db.execSQL(getUpdate1ColumnFromEq1Column(Keys.APP_START_LOAD_LISTS, startAppLoadLists, screenName));
	}

	private String getUpdate1ColumnFromEq1Column(Object updateKey, Object update, String screenName){
		return String.format("UPDATE accounts SET %s='%s' WHERE %s='%s'", updateKey, update, Keys.SCREEN_NAME, screenName);
	}

	public String[] getSelectListNames(String screenName){
		Cursor c = db.rawQuery(String.format("SELECT %s FROM accounts WHERE %s='%s'",
				Keys.SELECT_LIST_NAMES, Keys.SCREEN_NAME, screenName), null);
		c.moveToFirst();
		String[] result = c.getString(0).split(",");
		c.close();
		return result;
	}

	public String[] getNowStartAppLoadList(String screenName){
		Cursor c = db.rawQuery(String.format("SELECT %s FROM accounts WHERE %s='%s'",
				Keys.APP_START_LOAD_LISTS, Keys.SCREEN_NAME, screenName), null);
		c.moveToNext();
		String[] result = c.getString(0).split(",");
		c.close();
		return result;
	}
}

File: app/src/main/java/sugtao4423/lod/utils/DBUtil.kt
package sugtao4423.lod.utils

import android.content.Context
import android.database.Cursor
import sugtao4423.lod.Account
import sugtao4423.lod.AccountDB
import sugtao4423.lod.Keys

class DBUtil(context: Context) {

    private val db = AccountDB(context).writableDatabase

    fun dbClose() {
        db.close()
    }

    fun getAccount(screenName: String): Account? {
        val c = db.rawQuery(String.format("SELECT * FROM accounts WHERE %s = '%s'", Keys.SCREEN_NAME, screenName), null)
        if (!c.moveToFirst()) {
            return null
        }
        val account = getAccount(c)
        c.close()
        return account
    }

    fun getAccounts(): Array<Account> {
        val c = db.rawQuery("SELECT * FROM accounts", null)
        val accountList = arrayListOf<Account>()

        var mov = c.moveToFirst()
        while (mov) {
            accountList.add(getAccount(c))
            mov = c.moveToNext()
        }
        c.close()
        return accountList.toTypedArray()
    }

    private fun getAccount(c: Cursor): Account {
        c.apply {
            val screenName = getString(0)
            val ck = getString(1)
            val cs = getString(2)
            val at = getString(3)
            val ats = getString(4)
            val listAsTL = getLong(5)
            val autoLoadTLInterval = getInt(6)
            val listIds = getString(7)
            val listNames = getString(8)
            val appLoadLists = getString(9)

            val selectListIds = if (listIds == "") {
                LongArray(0)
            } else {
                val listIdsStr = listIds.split(Regex("\\s*,\\s*"))
                val ids = LongArray(listIdsStr.size)
                listIdsStr.mapIndexed { index, s ->
                    ids[index] = s.toLong()
                }
                ids
            }

            val selectListNames = if (listNames == "") arrayOf() else listNames.split(Regex("\\s*,\\s*")).toTypedArray()
            val startAppLoadLists = if (appLoadLists == "") arrayOf() else appLoadLists.split(Regex("\\s*,\\s*")).toTypedArray()

            return Account(screenName, ck, cs, at, ats, listAsTL, autoLoadTLInterval, selectListIds, selectListNames, startAppLoadLists)
        }
    }

    fun existsAccount(screenName: String): Boolean {
        val c = db.rawQuery(String.format("SELECT COUNT(*) FROM accounts WHERE %s = '%s'", Keys.SCREEN_NAME, screenName), null)
        c.moveToFirst()
        val count = c.getInt(0)
        c.close()
        return count > 0
    }

    fun addAccount(account: Account) {
        db.execSQL(String.format("INSERT INTO accounts VALUES('%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s')",
                account.screenName, account.consumerKey, account.consumerSecret,
                account.accessToken, account.accessTokenSecret, account.listAsTL, account.autoLoadTLInterval,
                account.selectListIds.joinToString(),
                account.selectListNames.joinToString(),
                account.startAppLoadLists.joinToString()))
    }

    fun deleteAccount(account: Account) {
        db.execSQL(String.format("DELETE FROM accounts WHERE %s='%s'", Keys.SCREEN_NAME, account.screenName))
    }

    fun updateListAsTL(listAsTL: Long, screenName: String) {
        db.execSQL(getUpdate1ColumnFromEq1Column(Keys.LIST_AS_TIMELINE, listAsTL, screenName))
    }

    fun updateAutoLoadTLInterval(autoLoadTLInterval: Int, screenName: String) {
        db.execSQL(getUpdate1ColumnFromEq1Column(Keys.AUTO_LOAD_TL_INTERVAL, autoLoadTLInterval, screenName))
    }

    fun updateSelectListIds(ids: String, screenName: String) {
        db.execSQL(getUpdate1ColumnFromEq1Column(Keys.SELECT_LIST_IDS, ids, screenName))
    }

    fun updateSelectListNames(names: String, screenName: String) {
        db.execSQL(getUpdate1ColumnFromEq1Column(Keys.SELECT_LIST_NAMES, names, screenName))
    }

    fun updateStartAppLoadLists(startAppLoadLists: String, screenName: String) {
        db.execSQL(getUpdate1ColumnFromEq1Column(Keys.APP_START_LOAD_LISTS, startAppLoadLists, screenName))
    }

    private fun getUpdate1ColumnFromEq1Column(updateKey: Any, update: Any, screenName: String): String {
        return String.format("UPDATE accounts SET %s = '%s' WHERE %s = '%s'", updateKey, update, Keys.SCREEN_NAME, screenName)
    }

    fun getSelectListNames(screenName: String): Array<String> {
        val c = db.rawQuery(String.format("SELECT %s FROM accounts WHERE %s = '%s'", Keys.SELECT_LIST_NAMES, Keys.SCREEN_NAME, screenName), null)
        c.moveToFirst()
        val result = c.getString(0).split(Regex("\\s*,\\s*"))
        c.close()
        return result.toTypedArray()
    }

    fun getNowStartAppLoadList(screenName: String): Array<String> {
        val c = db.rawQuery(String.format("SELECT %s FROM accounts WHERE %s = '%s'", Keys.APP_START_LOAD_LISTS, Keys.SCREEN_NAME, screenName), null)
        c.moveToNext()
        val result = c.getString(0).split(Regex("\\s*,\\s*"))
        c.close()
        return result.toTypedArray()
    }

}